### PR TITLE
Public API Refactor

### LINKED ISSUE #377: Shorter and Simpler Names across public api
Tentative Steps
1. Deprecate unused APIs
2. Rename APIs
3. Add new APIs
4. Add node/edge target into our API

TODO:
- how do we expect the user to read / query information for feature and targets? Does `DGBatch` need seperate attributes for e.g. `node_x_time` and `node_y_time`. I think not: just keep `node_time`. But then this requires that either
    1. User never knows which time is for feature, which is for target (not good)
    2. We must store an additional mask on `DGBatch` that users can use to sub-index the `node_time / edge_time / edge_index`

Current proposal:

**DGData**

```python
# Global Timestamps and event masks
time: Tensor  # [num_events]
edge_x_mask: Tensor  | None = None # [num_events]
edge_y_mask: Tensor  | None = None # [num_events]
node_x_mask: Tensor  | None = None # [num_events]
node_y_mask: Tensor  | None = None # [num_events]

# Edge events
edge_index: Tensor                            # [num_edge_events, 2]
edge_x: Tensor | None = None        # [num_edge_events, D_edge_x]
edge_y: Tensor | None = None        # [num_edge_events, D_edge_y]

# Node events
node_ids: Tensor | None                  # [num_node_events]
node_x: Tensor | None                     # [num_node_events, D_node_x]
node_y: Tensor | None                     # [num_node_events, D_node_y]

# Static entities
node_type: Tensor | None = None                # [num_nodes]
edge_type: Tensor | None = None               # [num_edges]
static_node_x: Tensor | None  = None        # [num_nodes, D_node_static]
```

**DGraph**
```python
  def start_time(self) -> Optional[int]:
  def end_time(self) -> Optional[int]:
  def num_nodes(self) -> int:
  def num_node_events(self) -> int:
  def num_edge_events(self) -> int:
  def num_events(self) -> int:
  def num_timestamps(self) -> int:
  
  def edge_index(self) -> Tensor:  
  def src(self) -> Tensor:
  def dst(self) -> Tensor:
  def edge_time(self) -> Tensor:
 
  def node_ids(self) -> Optional[Tensor]:
  def node_time(self) -> Optional[Tensor]:

  def node_x(self) -> Optional[Tensor]:
  def node_y(self) -> Optional[Tensor]:
  def edge_x(self) -> Optional[Tensor]:
  def edge_y(self) -> Optional[Tensor]:

  def node_type(self) -> Optional[Tensor]:
  def edge_type(self) -> Optional[Tensor]:
  def static_node_x(self) -> Optional[Tensor]:

  def node_x_dim(self) -> Optional[int]:
  def node_y_dim(self) -> Optional[int]:
  def edge_x_dim(self) -> Optional[int]:
  def edge_y_dim(self) -> Optional[int]:
  def static_node_x_dim(self) -> Optional[int]:
```

**DGBatch**
```python
class DGBatch:
  src: Tensor
  dst: Tensor
  edge_time: Tensor
  edge_x: Optional[Tensor] = None
  edge_type: Optional[Tensor] = None

  # edge target by mask
  edge_y_mask: Optional[Tensor] = None
  edge_y: Optional[Tensor] = None

  # node event feartures, i.e. tweets 
  node_x_ids: Optional[Tensor] = None
  node_x_time: Optional[Tensor] = None
  node_x: Optional[Tensor] = None

  # node target i.e. would user still subcribe to netflix next week, i.e. churn prediction, 0/1 label
  node_y_ids: Optional[Tensor] = None 
  node_y_time: Optional[Tensor] = None
  node_y: Optional[Tensor] = None
  
```

### PR DESCRIPTION
**Pull request by Jacob-Chmura**
Close #377

### CODE REVIEW

File: docs/tutorials/dgraph_tutorial.md
Comment (ntgbaoo) [inline docs/tutorials/dgraph_tutorial.md:1]: The definition of `DGData` [here](https://github.com/tgm-team/tgm/blob/dev/dgbatch_api_refactor/docs/quickstart.md#minimal-example) ned to be updated:
`timestamps` -> `time`

File: README.md
Comment (ntgbaoo) [inline README.md:1]: In the [Minimal Example](https://github.com/tgm-team/tgm/tree/dev/dgbatch_api_refactor?tab=readme-ov-file#minimal-example).
The line:
```
decoder = NodePredictor(in_dim=128, out_dim=train_dg.dynamic_node_feats_dim)
```
Need to be updated to:
```
decoder = NodePredictor(in_dim=128, out_dim=train_dg.node_x_dim)
```

File: examples/linkproppred/base3.py [hunk 1]
```diff
@@ -66,8 +66,8 @@ def eval(
             }
             perf_list.append(evaluator.eval(input_dict)[METRIC_TGB_LINKPROPPRED])
 
-        edgebank_model.update(batch.src, batch.dst, batch.time)
-        tcomem_model.update(batch.src, batch.dst, batch.time)
+        edgebank_model.update(batch.src, batch.dst, batch.edge_event_time)
```
Comment (shenyangHuang) [inline examples/linkproppred/base3.py]: `edge_event_time` with `edge_time`

File: docs/tutorials/dgraph_tutorial.md [hunk 2]
```diff
@@ -32,7 +32,7 @@ Because it's mutable, you can freely transform and prepare it before moving to t
 
 #### Features of `DGData`
 
-- Holds raw edge data (`edge_index`, `edge_timestamps`)
+- Holds raw edge data (`edge_index`, `edge_time`)
 - Holds *static node features*, *dynamic node features*, and *edge_features* (on CPU)
```
Comment (shenyangHuang) [inline docs/tutorials/dgraph_tutorial.md:36]: change name here too, `static_node_x` etc

File: docs/tutorials/dgraph_tutorial.md [hunk 3]
```diff
@@ -50,12 +50,12 @@ class DGData:
 
     Attributes:
         time_delta (TimeDeltaDG | str): Time granularity of the graph.
-        timestamps (Tensor): 1D tensor of all event timestamps [num_edge_events + num_node_events].
+        time (Tensor): 1D tensor of all event timestamps [num_edge_events + num_node_events].
         edge_mask (Tensor): Mask of edge events within `timestamps`.
         edge_index (Tensor): Edge connections [num_edge_events, 2].
```
Comment (shenyangHuang) [inline docs/tutorials/dgraph_tutorial.md:55]: stacking source and dst

File: docs/tutorials/hook_tutorial.md [hunk 4]
```diff
@@ -24,14 +24,14 @@ Hooks declare the following information
...
+| Hook Name                    | Type      | `requires` | `produces`                                | Description                                                                                          |
+| ---------------------------- | --------- | ---------- | ----------------------------------------- | ---------------------------------------------------------------------------------------------------- |
+| `NegativeEdgeSamplerHook`    | Stateless | None       | `neg`, `neg_time`                         | Generates random negatives for link prediction                                                       |
+| `TGBNegativeEdgeSamplerHook` | Stateless | None       | `neg`, `neg_time`, `neg_batch_list`       | Loads pre-computed negative edges for [TGB](https://tgb.complexdatalab.com/) datasets                |
+| `NeighborSamplerHook`        | Stateless | None       | `nbr_nids`, `nbr_edge_time`, `nbr_edge_x` | Uniform sampler neighbor for a given number of hops                                                  |
```
Comment (shenyangHuang) [inline docs/tutorials/hook_tutorial.md:31]: definitely improved from before, the old API of `nbr_feats` is not clear that this is edge level info

File: examples/linkproppred/tpnet.py [hunk 5]
```diff
@@ -125,43 +125,43 @@ def __init__(
...
+        edge_dst = batch.edge_dst
         neg = batch.neg
-        time = batch.time
+        time = batch.edge_time
         nbr_nids = batch.nbr_nids[0]
-        nbr_times = batch.nbr_times[0]
-        nbr_feats = batch.nbr_feats[0]
-        src_nbr_idx = batch.seed_node_nbr_mask['src']
-        dst_nbr_idx = batch.seed_node_nbr_mask['dst']
+        nbr_edge_time = batch.nbr_edge_time[0]
```
Comment (shenyangHuang) [inline examples/linkproppred/tpnet.py:133]: hmm, unrelated to the API but this naming here is confusing. Maybe should be `1hop_nbr_edge_time`

File: examples/nodeproppred/gcn.py [hunk 6]
```diff
@@ -100,12 +100,12 @@ def train(
 
     for batch in tqdm(loader):
         opt.zero_grad()
-        y_true = batch.dynamic_node_feats
+        y_true = batch.node_x
```
Comment (shenyangHuang) [inline examples/nodeproppred/gcn.py:103]: in the label PR, this becomes `node_y`

File: tgm/core/_storage/backends/array_backend.py [hunk 7]
```diff
@@ -48,7 +48,7 @@ def get_nodes(self, slice: DGSliceTracker) -> Set[int]:
             node_mask = (self._data.node_mask >= lb_idx) & (
                 self._data.node_mask < ub_idx
             )
-            node_event_nodes = self._data.node_ids[node_mask].unique().tolist()
+            node_event_nodes = self._data.node_x_nids[node_mask].unique().tolist()
```
Comment (shenyangHuang) [inline tgm/core/_storage/backends/array_backend.py]: we are keeping the `node_event_nodes` here?

File: tgm/data/dg_data.py [hunk 8]
```diff
@@ -218,8 +224,10 @@ def _maybe_cast_integral_tensor(x: Tensor, name: str) -> Tensor:
 
         # Validate static node features
         num_nodes = torch.max(self.edge_index).item() + 1  # 0-indexed
-        if self.node_ids is not None:
-            num_nodes = max(num_nodes, torch.max(self.node_ids).item() + 1)  # 0-indexed
+        if self.node_x_nids is not None:
```
Comment (shenyangHuang) [inline tgm/data/dg_data.py:227]: interestingly for `node_y_nids` in next PR, what if `max_node` exist in `node_y` what would that mean? 
to me it would just be invalid? you are predicting for an unknown node, it should be invalid

File: tgm/data/dg_data.py [hunk 9]
```diff
@@ -574,8 +582,8 @@ def from_csv(
             edge_time_col: Column name for edge times.
             edge_x_col: Optional edge feature columns.
             node_file_path: Optional CSV file for dynamic node features.
-            node_id_col: Column name for dynamic node event ids. Required if node_file_path is specified.
-            node_time_col: Column name for dynamic node event times. Required if node_file_path is specified.
+            node_x_nids_col: Column name for dynamic node event ids. Required if node_file_path is specified.
```
Comment (ntgbaoo) [inline tgm/data/dg_data.py]:
```suggestion
            node_x_nids_col: Column name for dynamic node event node ids. Required if node_file_path is specified.
```

File: tgm/data/dg_data.py [hunk 10]
```diff
@@ -706,8 +714,8 @@ def from_pandas(
             edge_time_col: Column name for edge times.
             edge_x_col: Optional edge feature columns.
             node_df: Optional DataFrame of dynamic node events.
-            node_id_col: Column name for dynamic node event ids. Required if node_file_path is specified.
-            node_time_col: Column name for dynamic node event times. Required if node_file_path is specified.
+            node_x_nids_col: Column name for dynamic node event ids. Required if node_file_path is specified.
```
Comment (ntgbaoo) [inline tgm/data/dg_data.py]:
```suggestion
            node_x_nids_col: Column name for dynamic node event node ids. Required if node_file_path is specified.
```

File: tgm/core/graph.py [hunk 11]
```diff
@@ -218,57 +218,80 @@ def _edges_cpu(self) -> Tuple[Tensor, Tensor, Tensor]:
...
+
+        If edge type exist, returns a tensor of shape (T x V x V).
+        """
+        edge_type = self._edge_type_cpu
+        if edge_type is not None:
+            edge_type = edge_type.to(self.device)
+        return edge_type
+
+    @_logged_cached_property
+    def _node_events_cpu(self) -> Tuple[Tensor, Tensor]:
```
Comment (ntgbaoo) [inline tgm/core/graph.py:269]:
```suggestion
    def _node_x_cpu(self) -> Tuple[Tensor, Tensor]:
```

File: tgm/core/graph.py [hunk 12]
```diff
@@ -218,57 +218,80 @@ def _edges_cpu(self) -> Tuple[Tensor, Tensor, Tensor]:
...
+        return self._storage.get_node_events(self._slice)
```
Comment (ntgbaoo) [inline tgm/core/graph.py:270]: Do we want to change implementation of storage to `get_node_x`? Eventually, I think we should since there will be `get_node_y`

File: tgm/core/_storage/backends/array_backend.py [hunk 13]
```diff
@@ -70,12 +70,12 @@ def get_edges(self, slice: DGSliceTracker) -> Tuple[Tensor, Tensor, Tensor]:
     def get_node_events(self, slice: DGSliceTracker) -> Tuple[Tensor, Tensor]:
```
Comment (ntgbaoo) [inline tgm/core/_storage/backends/array_backend.py:70]:
```suggestion
    def get_node_x(self, slice: DGSliceTracker) -> Tuple[Tensor, Tensor]:
```

File: tgm/hooks/hook_manager.py [hunk 14]
```diff
@@ -247,7 +247,15 @@ def _topological_sort_hooks(hooks: List[DGHook]) -> List[DGHook]:
...
-            ['src', 'dst', 'time', 'node_times', 'node_ids', 'edge_type', 'node_type']
+            [
+                'edge_src',
+                'edge_dst',
+                'edge_time',
+                'node_x_time',
+                'node_x_nids',
+                'edge_type',
+                'node_type',
+            ]
```
Comment (ntgbaoo) [inline tgm/hooks/hook_manager.py:258]: nit:
```suggestion
            [
                'edge_src',
                'edge_dst',
                'edge_time',
                'edge_type',
                'node_x_time',
                'node_x_nids',
                'node_type',
            ]
```

File: docs/quickstart.md
Comment (ntgbaoo) [inline docs/quickstart.md:1]: [This line](https://github.com/tgm-team/tgm/blob/dev/dgbatch_api_refactor/docs/quickstart.md?plain=1#L35) was missed.
Should be
```
edge_index = torch.stack([batch.edge_src, batch.edge_dst], dim=0)
```

File: README.md
Comment (ntgbaoo) [inline README.md:1]: [This line](https://github.com/tgm-team/tgm/blob/dev/dgbatch_api_refactor/README.md?plain=1#L130) was missed. Should be
```
edge_index = torch.stack([batch.edge_src, batch.edge_dst], dim=0)
```

**Review by benjaminnNgo — CHANGES_REQUESTED** [review]
Thanks @Jacob-Chmura for a very comprehensive PR. Took me a while to go over it lol.

File: examples/linkproppred/tpnet.py (hunk 5 above)
Comment (Jacob-Chmura) [reply examples/linkproppred/tpnet.py:133]: Yeah this would be a different (rather large) change. We can maybe do it in next release.

File: tgm/core/_storage/backends/array_backend.py (hunk 7 above)
Comment (Jacob-Chmura) [reply tgm/core/_storage/backends/array_backend.py]: I mean its just a temp var but I can update it

File: tgm/data/dg_data.py (hunk 8 above)
Comment (Jacob-Chmura) [reply tgm/data/dg_data.py:227]: Not sure about that. Seems like something that could easily happen?

File: tgm/core/graph.py (hunk 11 above)
Comment (Jacob-Chmura) [reply tgm/core/graph.py:269]: I think this conflicts too much with the backend call `get_node_x` which is getting dynamic node features

File: tgm/core/graph.py (hunk 12 above)
Comment (Jacob-Chmura) [reply tgm/core/graph.py:270]: hmm, I'll consider in the subsequent label PR. My only concern is that we already have `get_node_x` as the backend method to get dynamic node features. 

Arguably we can consolidate these all somehow.

File: docs/quickstart.md
Comment (Jacob-Chmura) [reply docs/quickstart.md:1]: Great catch

File: README.md
Comment (Jacob-Chmura) [reply README.md:1]: Great catch

File: tgm/data/dg_data.py (hunk 8 above)
Comment (shenyangHuang) [reply tgm/data/dg_data.py:227]: if the node doesn't exist in the graph then you can't predict for it? to discuss in next PR

**Review by benjaminnNgo — APPROVED** [review]
lgtm!
